ddl/partition.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ddl
+
+import (
+	"bytes"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/table"
+)
+
+var (
+	partitionMaxValue = "MAXVALUE"
+)
+
+// buildTablePartitionInfo builds partition info and checks for some errors.
+func buildTablePartitionInfo(ctx sessionctx.Context, d *ddl, s *ast.CreateTableStmt, cols []*table.Column) (*model.PartitionInfo, error) {
+	if s.Partition == nil {
+		return nil, nil
+	}
+	pi := &model.PartitionInfo{
+		Type:   s.Partition.Tp,
+		Enable: ctx.GetSessionVars().EnableTablePartition,
+	}
+	if s.Partition.Expr != nil {
+		buf := new(bytes.Buffer)
+		s.Partition.Expr.Format(buf)
+		pi.Expr = buf.String()
+		if s.Partition.Tp == model.PartitionTypeRange {
+			for _, col := range cols {
+				name := strings.Replace(col.Name.String(), ".", "`.`", -1)
+				if _, ok := s.Partition.Expr.(*ast.ColumnNameExpr); ok {
+					// TODO: check that the expression returns an integer.
+				}
+				if _, ok := s.Partition.Expr.(ast.ExprNode); ok {
+					if col.Tp != mysql.TypeLong && col.Tp != mysql.TypeLonglong && fmt.Sprintf("`%s`", name) == pi.Expr {
+						return nil, errors.Trace(ErrNotAllowedTypeInPartition.GenByArgs(pi.Expr))
+					}
+				}
+			}
+		}
+	} else if s.Partition.ColumnNames != nil {
+		pi.Columns = make([]model.CIStr, 0, len(s.Partition.ColumnNames))
+		for _, cn := range s.Partition.ColumnNames {
+			pi.Columns = append(pi.Columns, cn.Name)
+		}
+	}
+	for _, def := range s.Partition.Definitions {
+		// TODO: generate multiple global ID for paritions.
```

Comment:
Why ned to `generate multiple global ID`?

Reply:
Please @coocood answer the question.